added hl.import_genomicsdb

cseed opened the pull request:
I created and exported two genomicsdb shards using GATK.  The test compares genomicsdb import to importing the exported VCFs from GATK (modulo some minor differences).

This PR includes an 18MB compressed FASTA file for h38 chr20 (import_genomicsdb needs a FASTA file).  Github has a 100MB file size limit.  This seems fine, but if there is an objection we can switch to using git-lfs: https://git-lfs.github.com/

@lfrancioli I will test against gnomADv3 test data next and post some timings.

jigold commented on python/hail/methods/impex.py:
```diff
@@ -1714,6 +1714,59 @@ def import_vcf(path,
+           call_fields=oneof(str, sequenceof(str)),
+           reference_genome=nullable(reference_genome_type),
+           entry_field_array_elements_required=bool,
+           tar=nullable(str))
```
Is it easy for you to add `contig_recoding` like `import_vcf`? Otherwise, I'll add it in a separate PR once this goes in.

Also, the `entry_field_array_elements_required` is slightly different from `import_vcf`. Is there a way to unify them?

cseed commented on src/main/scala/is/hail/io/genomicsdb/ImportGenomicsDB.scala:
```diff
@@ -0,0 +1,231 @@
+    val rvb = new RegionValueBuilder(region)
+    val rv = RegionValue(region)
+
+    val it: java.util.Iterator[VariantContext] = gdbReader.iterator
```
Yeah, so I needed to do this to get it to compile.  I'm not quite sure why.  gdbReader.iterator is a CloseableTribbleIterator and that was somehow messing up the asScala monkey patch.

cseed commented on python/hail/methods/impex.py:
```diff
@@ -1714,6 +1714,59 @@ def import_vcf(path,
+    ----------
+    metadata_file : :obj:`str`
+        Path to GenomicsDB JSON metadata file.
+    call_fields : :obj:`list` of :obj:`str`
```
This would be a reasonable change.  I think the current setup is due to special treatment to GT in the VCF spec (it is clearly specified as a call, it must be first if present, etc.)  Can we leave it for a separate PR and I'll just make sure import_{vcf, genomicsdb} agree for now?

danking replied:
OK

cseed commented on build.gradle:
```diff
@@ -115,8 +115,7 @@ compileScala {
 
 dependencies {
-    //"compile group: 'com.intel', name: 'genomicsdb', version: '0.6.4-proto-3.0.0-beta-1'" currently does not support spark-core 2.11
-    compile files("libs/genomicsdb-0.6.4.jar")
```
It's in there, search for it.  (You might need to scroll down to get Github to load the entire page.)

cseed commented on src/main/scala/is/hail/io/genomicsdb/ImportGenomicsDB.scala:
```diff
@@ -0,0 +1,231 @@
+      .map { vc =>
+        region.clear()
+        rvb.start(typ.rvRowType)
+        reader.readRecord(vc, rvb, metadata.infoType, typ.entryType, dropSamples = false, metadata.canonicalFlags, metadata.infoFlagFieldNames)
```
At this point, I don't anticipate we're going to be using import_genomicsdb for anything except import to the native MatrixTable format, so I'm not inclined to invest in these optimizations without a clear use case.

cseed commented on python/hail/tests/test_methods.py:
```diff
@@ -1901,3 +1901,16 @@ def test_de_novo(self):
+    def test_import_genomicsdb(self):
+        rg = hl.ReferenceGenome.read('src/test/resources/grch38_chr20.json')
+        rg.add_sequence('src/test/resources/Homo_sapiens_assembly38_chr20.fasta.gz', 'src/test/resources/Homo_sapiens_assembly38_chr20.fasta.fai')
+        gdb = hl.import_genomicsdb('src/test/resources/gdb/test.json', reference_genome='GRCh38_chr20')
```
OK, I added these two tests, but I don't actually test with data which has missing array elements (the test data has a PGT field to test call_fields).

cseed commented:
OK, I think I've addressed all the comments.  I made some additional changes:

 - pipe input file directly into tar instead of writing to disk (writing to SSDs, I get ~100MB/s/core download saturating gs://),
 - report records read,
 - directly create OrderedRVD instead of coercing,
 - updated GenomicsDB to latest: 0.9.2-proto-3.0.0-beta-1+ed318f7e815 which involved revising GenomicsDBFeatureReader ctor call

danking commented on python/hail/tests/test_methods.py:
```diff
@@ -1976,12 +1976,22 @@ def test_window_by_locus(self):
+            gdb = hl.import_genomicsdb('src/test/resources/gdb/test.json',
+                                       reference_genome='GRCh38_chr20',
+                                       **kwargs)
+            gdb = gdb.annotate_entries(GT = hl.null(hl.tcall))
```
Does this test not check that we read the GTs correctly?

cseed replied:
Good catch.

The export files I compare against were created with GATK CombineGVCFs but as far as I can tell it won't output GTs (which itself is a special option for GenomicsDB).  I'll add another test to spot check a couple of GTs.
